File: app/src/main/java/com/example/cinemates/ui/fragment/BottomSheetFragment.java
package com.example.cinemates.ui.fragment;

import android.app.Dialog;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;

import com.example.cinemates.databinding.FragmentBottomsheetBinding;
import com.example.cinemates.model.Movie;
import com.example.cinemates.viewmodel.DbViewModel;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

/**
 * @author Antonio Di Nuzzo
 * Created 26/07/2022 at 09:01
 */
public class BottomSheetFragment extends BottomSheetDialogFragment {
    private FragmentBottomsheetBinding mBinding;
    private DbViewModel mViewModel;
    private Movie mMovie;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mViewModel = new ViewModelProvider(getActivity()).get(DbViewModel.class);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        BottomSheetDialog dialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);
        mBinding = FragmentBottomsheetBinding.inflate(getLayoutInflater());
        dialog.setContentView(mBinding.getRoot());
        mMovie = BottomSheetFragmentArgs.fromBundle(getArguments()).getMovie();
        Movie tmp = mViewModel.getMovie(mMovie);
        if (tmp != null)
            mMovie = tmp;

        mBinding.setMovie(mMovie);
        mBinding.favoriteBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // TODO: 26/07/2022 Add to favorite
                mMovie.setFavorite();
                update(dialog);
            }
        });

        mBinding.watchlistBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // TODO: 26/07/2022 Add to watchlist
                mMovie.setPersonalStatus(Movie.PersonalStatus.TO_SEE);
                update(dialog);
            }
        });

        mBinding.watchedBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // TODO: 26/07/2022 Add to watched
                mMovie.setPersonalStatus(Movie.PersonalStatus.SEEN);
                update(dialog);
            }
        });


        return dialog;
    }

    private void update(BottomSheetDialog dialog) {
        mViewModel.insert(mMovie);
        mBinding.setMovie(mMovie);
        dialog.dismiss();
    }


}

File: app/src/main/java/com/example/cinemates/view/ui/details/movie/BottomSheetFragment.kt
package com.example.cinemates.view.ui.details.movie

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.activityViewModels
import com.example.cinemates.databinding.FragmentBottomsheetBinding
import com.example.cinemates.model.data.PersonalStatus
import com.example.cinemates.view.viewmodel.DbMovieViewModel
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment

/**
 * @author Antonio Di Nuzzo
 * Created 26/07/2022 at 09:01
 */
class BottomSheetFragment : BottomSheetDialogFragment() {
    private var _binding: FragmentBottomsheetBinding? = null
    private val binding: FragmentBottomsheetBinding
        get() = _binding!!
    private val dbViewModel: DbMovieViewModel by activityViewModels()
    private val movieViewModel: MovieDetailsViewModel by activityViewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentBottomsheetBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
        _binding = FragmentBottomsheetBinding.inflate(layoutInflater)
        dialog.setContentView(binding.root)

        return dialog
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.apply {
            //listen selected movie changes for updating UI
            movieViewModel.selectedMovie.observe(viewLifecycleOwner) { movie ->
                setMovie(movie)
            }

            favoriteBtn.setOnClickListener {
                movieViewModel.setFavorite()
                val result = dbViewModel.setAsFavorite(movieViewModel.selectedMovie.value!!)

                makeResultToast(result, "Favorites")

                dismiss()
            }

            watchlistBtn.setOnClickListener {
                movieViewModel.setPersonalStatus(PersonalStatus.TO_SEE)
                val result = dbViewModel.updatePersonalStatus(movieViewModel.selectedMovie.value!!, PersonalStatus.TO_SEE)

                makeResultToast(result, "Watch list")

                dismiss()
            }

            watchedBtn.setOnClickListener {
                movieViewModel.setPersonalStatus(PersonalStatus.SEEN)
                val result = dbViewModel.updatePersonalStatus(movieViewModel.selectedMovie.value!!, PersonalStatus.SEEN)

                makeResultToast(result, "Watched list")

                dismiss()
            }


        }
    }

    private fun makeResultToast(result: Boolean, what: String) {
        val text = if (result) "Added to"
        else "Removed from"
        Toast.makeText(context, "$text $what", Toast.LENGTH_SHORT).show()
    }


    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}